### PR TITLE
Add tests for AWS object copy

### PR DESCRIPTION
**Pull request by dora-korpar**
Adds tests for objectCopy API with AWS backend. ~~Nothing had to be added to the API itself, but~~ a minor change was included because it directly impacted the testing. This is adding an error return on a GET call to AWS.
New features include 0-byte object support and copying objects across backends while retaining all other metadata.

### CODE REVIEW

File: lib/data/external/AwsClient.js [hunk 1]
```diff
@@ -32,7 +32,7 @@ class AwsClient {
             ContentLength: size,
         };
         if (keyContext.cipherBundle) {
-            uploadParams.ServerSideEncryption = 'aws:kms';
+            uploadParams.ServerSideEncryption = 'AES256';
```
Comment (LaurenSpiegel) [inline lib/data/external/AwsClient.js]: I agree with this change but why is it in this pr?

File: lib/data/external/AwsClient.js [hunk 2]
```diff
@@ -62,6 +62,7 @@ class AwsClient {
         const stream = request.createReadStream().on('error', err => {
             log.error('error streaming data from AWS', { error: err,
                 dataStoreName: this._dataStoreName });
+            return callback(err);
```
Comment (LaurenSpiegel) [inline lib/data/external/AwsClient.js]: why is this added? didn't we discuss that this callback is confusing since it is not actually called?  once the listener is put on the stream, the headers have already been sent back to the requesting client and we break the connection if there is an error.

File: tests/functional/aws-node-sdk/test/multipleBackend/objectCopy.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+            .catch(err => {
+                process.stdout.write(`Error in afterEach: ${err}\n`);
+                throw err;
+            });
+        });
+
+        it('should copy an object from mem to AWS', done => {
+            putSourceObj('mem', () => {
+                const copyParams = Object.assign({
+                    MetadataDirective: 'REPLACE',
```
Comment (LaurenSpiegel) [inline tests/functional/aws-node-sdk/test/multipleBackend/objectCopy.js]: can we test that the metadata has been properly copied?

File: tests/functional/aws-node-sdk/test/multipleBackend/put.js [hunk 4]
```diff
@@ -51,7 +51,7 @@ function awsGetCheck(objectKey, s3MD5, awsMD5, cb) {
             if (process.env.ENABLE_KMS_ENCRYPTION === 'true') {
                 // doesn't check ETag because it's different
                 // with every PUT with encryption
-                assert.strictEqual(res.ServerSideEncryption, 'aws:kms');
+                assert.strictEqual(res.ServerSideEncryption, 'AES256');
```
Comment (LaurenSpiegel) [inline tests/functional/aws-node-sdk/test/multipleBackend/put.js]: separate pr?

File: lib/data/external/AwsClient.js (hunk 2 above)
Comment (dora-korpar) [inline lib/data/external/AwsClient.js]: In objectCopy, data.get is called directly, so the get error isn't caught and the API continues with the data.put, which makes the server hangs indefinitely

**Comment by dora-korpar** [comment]
@ironman-machine go

**Comment by dora-korpar** [comment]
@ironman-machine go2

File: lib/data/external/AwsClient.js (hunk 2 above)
Comment (LaurenSpiegel) [inline lib/data/external/AwsClient.js]: ah, makes sense

File: tests/functional/aws-node-sdk/test/multipleBackend/objectCopy.js (hunk 3 above)
Comment (LaurenSpiegel) [inline tests/functional/aws-node-sdk/test/multipleBackend/objectCopy.js]: Thanks. As discussed, if someone is copying from location constraint x to location constraint y and the source object metadata has a location header (scal-location-constraint) and they have selected the "COPY" option for metadata, we should copy all of the metadata except for the location header (scal-location-constraint) which we should have as the actual destination location.

**Comment by dora-korpar** [comment]
@ironman-machine go3

File: lib/api/objectCopy.js [hunk 5]
```diff
@@ -84,6 +84,12 @@ function _prepMetadata(sourceObjMD, headers, sourceIsDestination, authInfo,
         });
         return { error: userMetadata };
     }
+    // If metadataDirective is 'COPY' and location constraint header is
+    // specified, new location constraint should override source object's
+    const locationHeader = 'x-amz-meta-scal-location-constraint';
```
Comment (LaurenSpiegel) [inline lib/api/objectCopy.js]: this is in the constants file already so can pull it from there

File: lib/api/objectCopy.js [hunk 6]
```diff
@@ -280,7 +286,9 @@ function objectCopy(authInfo, request, sourceBucket,
             // skip if source and dest the same or 0 byte object
             // still send along serverSideEncryption info so algo
             // and masterKeyId stored properly in metadata
-            if (sourceIsDestination || dataLocator.length === 0) {
+            if ((sourceIsDestination && !storeMetadataParams.metaHeaders
+            ['x-amz-meta-scal-location-constraint'])
```
Comment (LaurenSpiegel) [inline lib/api/objectCopy.js]: use the constant here

Comment (LaurenSpiegel) [inline lib/api/objectCopy.js]: also, isn't just checking whether this header is there not enough? shouldn't the check be that the specified location (if specified) is different from the data location of the source?  They could specify the location header but specify it to the same exact place.

File: tests/functional/aws-node-sdk/test/multipleBackend/objectCopy.js [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+const { createEncryptedBucketPromise } =
+    require('../../lib/utility/createEncryptedBucket');
+
+const awsLocation = 'aws-test';
+const bucket = 'buckettestmultiplebackendobjectcopy';
+const key = `somekey-${Date.now()}`;
+const copyKey = `copyKey-${Date.now()}`;
+const body = Buffer.from('I am a body', 'utf8');
+const correctMD5 = 'be747eb4b75517bf6b3cf7c5fbb62f3a';
+const locMetaHeader = 'scal-location-constraint';
```
Comment (LaurenSpiegel) [inline tests/functional/aws-node-sdk/test/multipleBackend/objectCopy.js]: use the constant with the x-amz-meta sliced off

**Comment by dora-korpar** [comment]
@ironman-machine go

**Comment by dora-korpar** [comment]
@ironman-machine gogo
